src/main/java/com/gitblit/manager/FilestoreManager.java

```diff
@@ -0,0 +1,368 @@
+
+		//Handle access control
+		if (!user.canPush(repo)) {
+			if (user == UserModel.ANONYMOUS) {
+				return Status.AuthenticationRequired;
+			} else {
+				return Status.Error_Unauthorized;
+			}
+		}
+		
+		//Handle object details
+		if (!isValidOid(oid)) { return Status.Error_Invalid_Oid; }
+		
+		if (fileCache.containsKey(oid)) {
+			FilestoreModel item = fileCache.get(oid);
+			
+			if (!item.isInErrorState() && (size != UNDEFINED_SIZE) && (item.getSize() != size)) {
+				return Status.Error_Size_Mismatch;
+			}
+			
+			if (!item.repositories.contains(repo)) {
+				item.repositories.add(repo.name);
+			}
+			
+			if (item.isInErrorState()) {
+				item.reset(user, size);
+			}
+		} else {
+			
+			if (size  < 0) {return Status.Error_Invalid_Size; }
+			if ((getMaxUploadSize() != UNDEFINED_SIZE) && (size > getMaxUploadSize())) { return Status.Error_Exceeds_Size_Limit; }
+			
+			fileCache.put(oid, new FilestoreModel(oid, size, user, repo.name));
+		}
+		
+		return fileCache.get(oid).getStatus();
+	}
+
+	@Override
+	public FilestoreModel.Status uploadBlob(String oid, long size, UserModel user, RepositoryModel repo, InputStream streamIn) {
+
+		//Access control and object logic
+		Status state = addObject(oid, size, user, repo); 
+		
+		if (state != Status.Upload_Pending) { 
+			return state;
+		}
+		
+		FilestoreModel item = fileCache.get(oid);
+		
+		if (!item.actionUpload(user)) {
+			return Status.Upload_In_Progress;
+		} else {
+			long actualSize = 0;
+			File file = getStoragePath(oid);
+
+			try {
+				file.getParentFile().mkdirs();
+				file.createNewFile();
+				
+				FileOutputStream streamOut = new FileOutputStream(file);
+				
+				actualSize = IOUtils.copyLarge(streamIn, streamOut);
+				
+				streamOut.flush();
+				streamOut.close();
+				
+				if (item.getSize() != actualSize) {
+					item.setStatus(Status.Error_Size_Mismatch, user);
+					
+					logger.warn(MessageFormat.format("Failed to upload blob {0} due to size mismatch, expected {1} got {2}", 
+							oid, item.getSize(), actualSize));
+				} else {
+					
+					FileInputStream fileForHash = new FileInputStream(file);
+					String actualOid = DigestUtils.sha256Hex(fileForHash);
+					fileForHash.close();
+					
+					if (oid.equalsIgnoreCase(actualOid)) {
+						item.setStatus(Status.Available, user);
+					} else {
+						item.setStatus(Status.Error_Hash_Mismatch, user);
+						
+						logger.warn(MessageFormat.format("Failed to upload blob {0} due to hash mismatch, got {1}", oid, actualOid));
+					}
+				}
+			} catch (Exception e) {
+				
+				item.setStatus(Status.Error_Unknown, user);
+				logger.warn(MessageFormat.format("Failed to upload blob {0}", oid), e);
+			}
+			
+			if (item.isInErrorState()) {
+				file.delete();
+				item.repositories.remove(repo.name);
+			}
+		}
+		
+		return item.getStatus();
+	}
+	
+	private FilestoreModel.Status canGetObject(String oid, UserModel user, RepositoryModel repo) {
+	
+		//Access Control
+		if (!user.canView(repo)) {
```

Comment:
Ah yes I had this thought a few times.  I was going to put ``canClone()`` but then in the case that a user has view permissions but not clone permissions they would be unable to request the blob via the web interface (not sure how likely that is).  Probably a more realistic scenario though is for in a later PR I'm hoping to use the filestore to handle ticket attachments.  In that case I thought it may be more feasible to have someone with view but not clone rights, what do you think?